File: api/sync.js
// アプリの監視銘柄リスト（hold/trade方針・追加/削除）を config.json に反映する（任意機能）。
// 必要な環境変数（Vercelプロジェクト設定）:
//   GH_TOKEN : GitHub Personal Access Token（repo / contents:write 権限）
//   GH_REPO  : "kzw-skcs/invest-alert-pwa"
// リクエストボディ: { "stocks": [ {ticker,name,sector,tradePolicy,stooq?,yahoo?}, ... ] }
const API = "https://api.github.com";

export default async function handler(req, res) {
  if (req.method !== "POST") return res.status(405).json({ error: "POST only" });
  const token = process.env.GH_TOKEN, repo = process.env.GH_REPO;
  if (!token || !repo) return res.status(501).json({ error: "サーバー未設定（GH_TOKEN/GH_REPO）" });
  try {
    const body = typeof req.body === "string" ? JSON.parse(req.body) : req.body;
    const stocks = body && body.stocks;
    if (!Array.isArray(stocks) || !stocks.length)
      return res.status(400).json({ error: "stocks配列が必要です" });

    // 正規化（危険な値を除去し、必要な補完）
    const clean = [];
    const seen = new Set();
    for (const s of stocks) {
      let t = (s.ticker || "").toString().trim().toUpperCase();
      if (!t || !/^[A-Z0-9.\-]{1,10}$/.test(t) || seen.has(t)) continue;
      seen.add(t);
      const pol = (s.tradePolicy === "trade") ? "trade" : "hold";
      clean.push({
        ticker: t,
        stooq: (s.stooq || (t.toLowerCase() + ".us")),
        name: (s.name || t).toString().slice(0, 60),
        sector: (s.sector || "その他").toString().slice(0, 20),
        tradePolicy: pol,
        ...(s.yahoo ? { yahoo: s.yahoo } : {})
      });
    }
    if (!clean.length) return res.status(400).json({ error: "有効な銘柄がありません" });

    const h = { Authorization: `Bearer ${token}`, Accept: "application/vnd.github+json" };
    const path = "config.json";
    const cur = await fetch(`${API}/repos/${repo}/contents/${path}`, { headers: h });
    if (cur.status !== 200) return res.status(500).json({ error: "config.json取得失敗" });
    const j = await cur.json();
    const cfg = JSON.parse(Buffer.from(j.content, "base64").toString("utf-8"));
    cfg.stocks = clean;
    const content = Buffer.from(JSON.stringify(cfg, null, 2), "utf-8").toString("base64");
    const put = await fetch(`${API}/repos/${repo}/contents/${path}`, {
      method: "PUT", headers: h,
      body: JSON.stringify({ message: "update watchlist from app", content, sha: j.sha })
    });
    if (!put.ok) return res.status(500).json({ error: "commit失敗", detail: await put.text() });
    return res.status(200).json({ ok: true, count: clean.length });
  } catch (e) {
    return res.status(500).json({ error: String(e) });
  }
}

File: api/sync.py
# -*- coding: utf-8 -*-
"""POST /api/sync — アプリの銘柄リスト変更(追加/hold⇔trade/非表示)をconfig.jsonへ反映。
body: {"stocks": [{"ticker","name","sector","subSector","tradePolicy","conviction","hidden"}...]}"""
import json
from http.server import BaseHTTPRequestHandler
from _github import get_file, put_file


class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            length = int(self.headers.get("Content-Length", 0))
            body = json.loads(self.rfile.read(length) or b"{}")
            incoming = {s["ticker"]: s for s in body.get("stocks", [])}
            cfg, sha = get_file("config.json")
            existing = {s["ticker"]: s for s in cfg["stocks"]}
            merged = []
            for t, s in incoming.items():
                base = existing.get(t, {})
                base.update({
                    "ticker": t,
                    "name": s.get("name", base.get("name", t)),
                    "sector": s.get("sector", base.get("sector", "その他")),
                    "subSector": s.get("subSector", base.get("subSector", "その他モート")),
                    "tradePolicy": s.get("tradePolicy", base.get("tradePolicy", "hold")),
                    "conviction": s.get("conviction", base.get("conviction", 3)),
                })
                base.setdefault("stooq", t.lower() + ".us")
                if not s.get("hidden"):
                    merged.append(base)
            cfg["stocks"] = merged
            put_file("config.json", cfg, sha, "app: 銘柄リスト同期")
            self._json(200, {"ok": True, "count": len(merged)})
        except Exception as e:
            self._json(500, {"ok": False, "error": str(e)})

    def _json(self, code, obj):
        b = json.dumps(obj, ensure_ascii=False).encode()
        self.send_response(code)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(b)))
        self.end_headers()
        self.wfile.write(b)
